File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/GainBlockRandomMonsterAction.ts
import Async from "../../../../../Core/Async";
import { AttackEffect } from "../../DataDefine/AttackEffect";
import DungeonManager from "../../DungeonManager";
import { FlashAtkImgEffect } from "../../Effect/FlashAtkImgEffect";
import AbstractCreature from "../../Unit/AbstractCreature";
import AbstractGameAction, { ActionType } from "../AbstractGameAction";

export class GainBlockRandomMonsterAction extends AbstractGameAction{
    Type = ActionType.BLOCK
    public constructor(public Source:AbstractCreature,public Amount:number) {
        super();
    }
    
    public async Update(): Promise<boolean> {
        if(this.Duration == 0){
            let target = DungeonManager.Inst.CurrentRoom.Monsters.Random();
            if(target){
                DungeonManager.EffectManager.Play(new FlashAtkImgEffect(target.X,target.Y,AttackEffect.SHIELD));
                target.Block += this.Amount;
            }
        }
        await Async.Delay(500);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Combat/GainBlockRandomMonsterAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Event;

namespace ModTheSpire.Effect.Other
{
    public class GainBlockRandomMonsterAction : IGameAction
    {
        private readonly int mBlockAmount;
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.BLOCK;
        public void Init()
        {
        }

        public GainBlockRandomMonsterAction(int blockAmount)
        {
            mBlockAmount = blockAmount;
        }
        
        public async UniTask<bool> Update()
        {
            var target = DungeonManager.Inst.CurrentRoom.Monsters.Random();
            if(target != null){
                DungeonManager.EffectManager.Play(new FlashAtkImgEffect(target.DisplayObject.X,target.DisplayObject.Y,AttackEffect.SHIELD));
                target.Block += this.mBlockAmount;
            }
            await UniTask.Delay(500);
            return true;
        }
    }
}